subsys/lorawan/services/clock_sync.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2022 Martin Jäger <martin@libre.solar>
+ * Copyright (c) 2022 tado GmbH
+ *
+ * Parts of this implementation were inspired by LmhpClockSync.c from the
+ * LoRaMac-node firmware repository https://github.com/Lora-net/LoRaMac-node
+ * written by Miguel Luis (Semtech).
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "lorawan_services.h"
+
+#include <LoRaMac.h>
+#include <zephyr/kernel.h>
+#include <zephyr/lorawan/lorawan.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/random/rand32.h>
+
+LOG_MODULE_REGISTER(lorawan_clock_sync, CONFIG_LORAWAN_SERVICES_LOG_LEVEL);
+
+/**
+ * Version of LoRaWAN Application Layer Clock Synchronization Specification
+ *
+ * This implementation only supports TS003-2.0.0, as the previous revision TS003-1.0.0
+ * requested to temporarily disable ADR and and set nb_trans to 1. This causes issues on the
+ * server side and is not recommended anymore.
+ */
+#define CLOCK_SYNC_PACKAGE_VERSION 2
+
+/* Maximum length of clock sync answers */
+#define MAX_CLOCK_SYNC_ANS_LEN 6
+
+/* Delay between consecutive transmissions of AppTimeReq */
+#define CLOCK_RESYNC_DELAY 10
+
+enum clock_sync_commands {
+	CLOCK_SYNC_CMD_PKG_VERSION                 = 0x00,
+	CLOCK_SYNC_CMD_APP_TIME                    = 0x01,
+	CLOCK_SYNC_CMD_DEVICE_APP_TIME_PERIODICITY = 0x02,
+	CLOCK_SYNC_CMD_FORCE_DEVICE_RESYNC         = 0x03,
+};
+
+struct clock_sync_context {
+	/** Work item for regular (re-)sync requests (uplink messages) */
+	struct k_work_delayable resync_work;
+	/** Continuously incremented token to map clock sync answers and requests */
+	uint8_t req_token;
+	/** Number of requested clock sync requests left to be transmitted */
+	uint8_t nb_transmissions;
+	/**
+	 * Offset to be added to system uptime to get GPS time (as used by LoRaWAN)
+	 */
+	int64_t time_correction;
+	/**
+	 * AppTimeReq retransmission interval in seconds
+	 *
+	 * Valid range between 128 (0x80) and 8388608 (0x800000)
+	 */
+	uint32_t periodicity;
+	/** Indication if at least one valid time correction was received */
+	bool synchronized;
+};
+
+static struct clock_sync_context ctx;
+
+static struct k_work_q *workq;
+
+/**
+ * Writes the DeviceTime into the buffer.
+ *
+ * @returns number of bytes written or -ENOSPC in case of error
+ */
+static int clock_sync_serialize_device_time(uint8_t *buf, size_t size)
+{
+	uint64_t device_time = k_uptime_get() / 1000 + ctx.time_correction;
+
+	if (size < 4) {
+		return -ENOSPC;
+	}
+
+	buf[0] = (device_time >> 0) & 0xFF;
+	buf[1] = (device_time >> 8) & 0xFF;
+	buf[2] = (device_time >> 16) & 0xFF;
+	buf[3] = (device_time >> 24) & 0xFF;
+
+	return 4;
```

Comment:
would be nicer to see `sizeof(device_time)` (if it can be `int32_t`) ... oh the joys of manual bitstuffing / serialization / deserialization